coderd/database/querier_test.go

```diff
@@ -3587,6 +3588,752 @@ func TestOrganizationDeleteTrigger(t *testing.T) {
+			require.Equal(t, backoff.PresetID, tmpl1V1.preset.ID)
+			require.Equal(t, int32(2), backoff.NumFailed)
+		}
+	})
+
+	t.Run("check LastBuildAt timestamp", func(t *testing.T) {
+		t.Parallel()
+
+		db, _ := dbtestutil.NewDB(t)
+		ctx := testutil.Context(t, testutil.WaitShort)
+		dbgen.Organization(t, db, database.Organization{
+			ID: orgID,
+		})
+		dbgen.User(t, db, database.User{
+			ID: userID,
+		})
+		lookbackPeriod := time.Hour
+
+		tmpl1 := createTemplate(t, db, orgID, userID)
+		tmpl1V1 := createTmplVersion(t, db, tmpl1, tmpl1.ActiveVersionID, now, &tmplVersionOpts{
+			DesiredInstances: 6,
+		})
+		createWorkspaceBuild(t, ctx, db, tmpl1, tmpl1V1, orgID, now, &workspaceBuildOpts{
+			successfulJob: false,
+			createdAt:     now.Add(-lookbackPeriod - time.Minute), // earlier than lookback period - skipped
+		})
+		createWorkspaceBuild(t, ctx, db, tmpl1, tmpl1V1, orgID, now, &workspaceBuildOpts{
+			successfulJob: false,
+			createdAt:     now.Add(-4 * time.Minute),
+		})
+		createWorkspaceBuild(t, ctx, db, tmpl1, tmpl1V1, orgID, now, &workspaceBuildOpts{
+			successfulJob: false,
+			createdAt:     now.Add(-0 * time.Minute),
+		})
+		createWorkspaceBuild(t, ctx, db, tmpl1, tmpl1V1, orgID, now, &workspaceBuildOpts{
+			successfulJob: false,
+			createdAt:     now.Add(-3 * time.Minute),
+		})
+		createWorkspaceBuild(t, ctx, db, tmpl1, tmpl1V1, orgID, now, &workspaceBuildOpts{
+			successfulJob: false,
+			createdAt:     now.Add(-1 * time.Minute),
+		})
+		createWorkspaceBuild(t, ctx, db, tmpl1, tmpl1V1, orgID, now, &workspaceBuildOpts{
+			successfulJob: false,
+			createdAt:     now.Add(-2 * time.Minute),
+		})
+
+		backoffs, err := db.GetPresetsBackoff(ctx, now.Add(-lookbackPeriod))
+		require.NoError(t, err)
+
+		require.Len(t, backoffs, 1)
+		{
+			backoff := backoffs[0]
+			require.Equal(t, backoff.TemplateVersionID, tmpl1.ActiveVersionID)
+			require.Equal(t, backoff.PresetID, tmpl1V1.preset.ID)
+			require.Equal(t, int32(5), backoff.NumFailed)
+			// make sure LastBuildAt is equal to latest failed build timestamp
+			require.Equal(t, 0, now.Compare(backoff.LastBuildAt))
+		}
+	})
+
+	t.Run("failed job outside lookback period", func(t *testing.T) {
+		t.Parallel()
+
+		db, _ := dbtestutil.NewDB(t)
+		ctx := testutil.Context(t, testutil.WaitShort)
+		dbgen.Organization(t, db, database.Organization{
+			ID: orgID,
+		})
+		dbgen.User(t, db, database.User{
+			ID: userID,
+		})
+		lookbackPeriod := time.Hour
+
+		tmpl1 := createTemplate(t, db, orgID, userID)
+		tmpl1V1 := createTmplVersion(t, db, tmpl1, tmpl1.ActiveVersionID, now, &tmplVersionOpts{
+			DesiredInstances: 1,
+		})
+
+		createWorkspaceBuild(t, ctx, db, tmpl1, tmpl1V1, orgID, now, &workspaceBuildOpts{
+			successfulJob: false,
+			createdAt:     now.Add(-lookbackPeriod - time.Minute), // earlier than lookback period - skipped
+		})
+
+		backoffs, err := db.GetPresetsBackoff(ctx, now.Add(-lookbackPeriod))
+		require.NoError(t, err)
+		require.Len(t, backoffs, 0)
+	})
```

Comment:
This is a nice, thorough set of tests and I appreciate it.

Reply:
@spikecurtis actually I'm only testing:
- `GetPresetsBackoff` SQL query
- multi agent workspaces for `workspace_prebuilds` view
- maybe some other things are tested indirectly
<hr/>

Mostly we're still relying on higher-level tests, like: reconciliation loop tests, etc.